pkg/discord/main.go

```diff
@@ -771,6 +795,145 @@ func fetchFile(data shared.CommandData) (*models.File, error) {
 	}
 }
 
+// Check common time formats
+func parseTime(timeStr string) (time.Time, error) {
+	now := time.Now()
+	formats := []string{
+		// Time only
+		"3:04PM",
+		"3:04:05PM",
+		"15:04",
+		"15:04:05",
+		// Date & time
+		"Jan 2 3:04PM",
+		"Jan 2 3:04:05PM",
+		"Jan 2 15:04",
+		"Jan 2 15:04:05",
+		"1/2 3:04PM",
+		"1/2 3:04:05PM",
+		"1/2 15:04",
+		"1/2 15:04:05",
+	}
+	var outTime time.Time
+	var err error
+	for i, format := range formats {
+		if t, err := time.Parse(format, timeStr); err == nil {
+			if i <= 3 {
+				outTime = time.Date(now.Year(), now.Month(), now.Day(), t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), now.Location())
+			} else {
+				outTime = time.Date(now.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), now.Location())
+			}
+			fmt.Println(time.Now().Sub(now))
+			return outTime, nil
+		}
+	}
+	return time.Time{}, err
+}
+
+func checkIn(data shared.CommandData) []shared.FunctionResponse {
+	if len(data.Args) == 0 {
+		return startSession(data, time.Now())
+	} else if inTime, err := parseTime(strings.Join(data.Args[0:], " ")); err != nil {
+		return []shared.FunctionResponse{
+			{
+				ChannelID: data.ChannelID,
+				Error:     "Failed to parse provided time. Try formatting your time as `3:04PM` or `Jan 2 3:04PM` if you're starting a session from another day.",
+			},
+		}
+	} else {
+		return startSession(data, inTime)
+	}
+}
+
+func startSession(data shared.CommandData, inTime time.Time) []shared.FunctionResponse {
+	session := models.VolunteerSession{
+		BrigadeID:     data.Brigade.ID,
+		DiscordUserID: data.Author.ID,
+		StartTime:     inTime,
+	}
+	if inTime.After(time.Now()) {
+		return []shared.FunctionResponse{
+			{
+				ChannelID: data.ChannelID,
+				Error:     "You can't start a volunteer session in the future. Try an earlier time.",
+			},
+		}
+	}
+	if err := db.Create(&session).Error; err != nil {
+		return []shared.FunctionResponse{
+			{
+				ChannelID: data.ChannelID,
+				Error:     "Failed to start volunteer session. Try again later.",
+			},
+		}
+	}
+	return []shared.FunctionResponse{
+		{
+			ChannelID: data.ChannelID,
+			Success:   "Volunteer session for <@!" + data.Author.ID + "> started. Use `!out` to end your session.",
+		},
+	}
+}
+
+func checkOut(data shared.CommandData) []shared.FunctionResponse {
+	if len(data.Args) == 0 {
+		return endSession(data, time.Now())
+	}
+	if outTime, err := parseTime(strings.Join(data.Args[0:], " ")); err != nil {
+		return []shared.FunctionResponse{
+			{
+				ChannelID: data.ChannelID,
+				Error:     "Failed to parse provided time. Try formatting your time as `3:04PM` or `Jan 2 3:04PM` if you're ending a session from another day.",
+			},
+		}
+	} else {
+		return endSession(data, outTime)
+	}
+}
+
+func endSession(data shared.CommandData, outTime time.Time) []shared.FunctionResponse {
```

Comment:
An issue may happen if multiple !checkIns are called and then attempted to close. Perhaps have a limit on 'active' sessions i.e. 1

Reply:
Good idea, will add a check to the startSession for if an existing session is going.